pkg/updatechecker/updatechecker.go

```diff
@@ -307,51 +283,142 @@ func CheckForUpdates(opts CheckForUpdatesOpts) (int64, error) {
-			if err != nil {
-				logger.Error(errors.Wrap(err, "failed to get update downstream status"))
-				continue
-			}
+	return availableUpdates, nil
+}
 
-			if status == storetypes.VersionPendingConfig {
-				logger.Infof("not deploying version %d because it's %s", sequence, status)
-				continue
-			}
+type DownloadUpdateOptions struct {
+	AppID          string
+	ClusterID      string
+	Update         upstreamtypes.Update
+	SkipPreflights bool
+	Deploy         bool
+	IsCLI          bool
+}
 
-			if err := version.DeployVersion(a.ID, sequence); err != nil {
-				logger.Error(errors.Wrap(err, "failed to queue update for deployment"))
-			}
+func downloadUpdate(opts DownloadUpdateOptions) error {
+	baseArchiveDir, err := GetBaseArchiveDirForVersion(opts.AppID, opts.ClusterID, opts.Update.VersionLabel)
+	if err != nil {
+		return errors.Wrapf(err, "failed to get base archive dir for version %s", opts.Update.VersionLabel)
+	}
+	defer os.RemoveAll(baseArchiveDir)
 
-			// preflights reporting
-			go func() {
-				err = reporting.ReportAppInfo(a.ID, sequence, opts.SkipPreflights, opts.IsCLI)
-				if err != nil {
-					logger.Debugf("failed to update preflights reports: %v", err)
-				}
-			}()
+	sequence, err := upstream.DownloadUpdate(opts.AppID, baseArchiveDir, opts.Update.Cursor, opts.SkipPreflights)
+	if err != nil {
+		return errors.Wrap(err, "failed to download update")
+	}
+
+	if !opts.Deploy {
+		return nil
+	}
+
+	status, err := store.GetStore().GetStatusForVersion(opts.AppID, opts.ClusterID, sequence)
+	if err != nil {
+		return errors.Wrap(err, "failed to get update downstream status")
+	}
+
+	if status == storetypes.VersionPendingConfig {
+		logger.Infof("not deploying version %d because it's %s", sequence, status)
+		return nil
+	}
+
+	if err := version.DeployVersion(opts.AppID, sequence); err != nil {
+		logger.Error(errors.Wrap(err, "failed to queue update for deployment"))
+	}
+
+	// preflights reporting
+	go func() {
+		err = reporting.ReportAppInfo(opts.AppID, sequence, opts.SkipPreflights, opts.IsCLI)
+		if err != nil {
+			logger.Debugf("failed to update preflights reports: %v", err)
 		}
 	}()
 
-	return availableUpdates, nil
+	return nil
+}
+
+// GetBaseArchiveDirForVersion returns the base archive directory for a given version label.
+// the base archive directory contains data such as config values.
+// caller is responsible for cleaning up the created archive dir.
+func GetBaseArchiveDirForVersion(appID string, clusterID string, targetVersionLabel string) (string, error) {
+	appVersions, err := store.GetStore().GetAppVersions(appID, clusterID)
+	if err != nil {
+		return "", errors.Wrapf(err, "failed to get app versions for app %s", appID)
+	}
+	if len(appVersions.AllVersions) == 0 {
+		return "", errors.Errorf("no app versions found for app %s in downstream %s", appID, clusterID)
+	}
+
+	mockVersion := &downstreamtypes.DownstreamVersion{
+		Sequence: -1, // to id the mocked version and be able to retrieve it later
```

Comment:
If there is no semver, this will be placed at the bottom of the list.  How about `math.MaxInt64`?